Dev/ltc2983

jemfgeronimo opened the pull request:
## Pull Request Description

The LTC2983 measures a wide variety of temperature sensors and digitally outputs the result, in °C or °F, with 0.1°C accuracy and 0.001°C resolution. The LTC2983 can measure the temperature of virtually all standard (type B, E, J, K, N, S, R, T) or custom thermocouples, automatically compensate for cold junction temperatures and linearize the results. The device can also measure temperature with standard 2-, 3- or 4-wire RTDs, thermistors and diodes. It has 20 reconfigurable analog inputs enabling many sensor connections and configuration options. The LTC2983 includes excitation current sources and fault detection circuitry appropriate for each type of temperature sensor.

## PR Type
- [ ] Bug fix (change that fixes an issue)
- [x] New feature (change that adds new functionality)
- [ ] Breaking change (has dependencies in other repos or will cause CI to fail)

## PR Checklist
- [x] I have followed the [Coding style guidelines](http://analogdevicesinc.github.io/no-OS/drivers_guide.html#coding-style)
- [x] I have performed a self-review of the changes
- [x] I have commented my code, at least hard-to-understand parts
- [ ] I have build all projects affected by the changes in this PR
- [x] I have tested in hardware affected projects, at the relevant boards
- [x] I have signed off all commits from this PR
- [ ] I have updated the documentation (wiki pages, ReadMe etc), if applies

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/analogdevicesinc/no-OS?pullRequest=2300) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/analogdevicesinc/no-OS?pullRequest=2300) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/analogdevicesinc/no-OS?pullRequest=2300) it.</sub>

CiprianRegus commented on drivers/temperature/ltc2983/ltc2983.c:
```diff
@@ -0,0 +1,644 @@
+	if (ret)
+		goto free_err;
+
+	if (init_param->dev_type == ID_LTC2983)
+		descriptor->max_channels_nr = 20;
+	else if (init_param->dev_type == ID_LTC2984)
+		descriptor->max_channels_nr = 20;
+	else if (init_param->dev_type == ID_LTC2986)
+		descriptor->max_channels_nr = 10;
```
I'd say that a `switch` statement would fit better

CiprianRegus commented on drivers/temperature/ltc2983/ltc2983.c:
```diff
@@ -0,0 +1,644 @@
+	if (!descriptor->num_channels) // should be at least one channel
+		goto spi_err;
+
+	ret = no_os_gpio_get(&descriptor->gpio_rstn, &init_param->gpio_rstn);
```
Can the LTC2983 be software reset? In that case, we can make the reset GPIO optional

CiprianRegus commented on drivers/temperature/ltc2983/ltc2983.c:
```diff
@@ -0,0 +1,644 @@
+
+	ret = ltc2983_setup(descriptor);
+	if (ret)
+		return ret;
```
`goto gpio_err`

CiprianRegus commented on drivers/temperature/ltc2983/ltc2983.c:
```diff
@@ -0,0 +1,644 @@
+	if (ret)
+		return ret;
+
+	data = (data & ~mask) | (val & mask);
```
For other drivers, we usually assume the `val` is actually the value of the masked field, without any shifting. I think we should do the same here, since it's more straight forward for the function caller.

CiprianRegus commented on drivers/temperature/ltc2983/ltc2983.c:
```diff
@@ -0,0 +1,644 @@
+		if (!device->sensors[i])
+			continue;
+
+		if (device->sensors[i]->type <= LTC2983_THERMOCOUPLE_CUSTOM)
+			ltc2983_thermocouple_assign_chan(device,
+							 device->sensors[i]);
+		else if (device->sensors[i]->type <= LTC2983_RTD_CUSTOM)
+			ltc2983_rtd_assign_chan(device, device->sensors[i]);
+		else if (device->sensors[i]->type <= LTC2983_THERMISTOR_CUSTOM)
+			ltc2983_thermistor_assign_chan(device,
+						       device->sensors[i]);
+		else if (device->sensors[i]->type == LTC2983_DIODE)
+			ltc2983_diode_assign_chan(device, device->sensors[i]);
+		else if (device->sensors[i]->type == LTC2983_RSENSE)
+			ltc2983_r_sense_assign_chan(device, device->sensors[i]);
+		else if (device->sensors[i]->type == LTC2983_DIRECT_ADC)
+			ltc2983_adc_assign_chan(device, device->sensors[i]);
+		else if (device->sensors[i]->type == LTC2983_ACTIVE_TEMP)
+			ltc2983_temp_assign_chan(device, device->sensors[i]);
```
`switch`

CiprianRegus commented on drivers/temperature/ltc2983/ltc2983.c:
```diff
@@ -0,0 +1,644 @@
+ * @param val - channel data / temperature
+ * @return 0 in case of success, errno errors otherwise
+ */
+int ltc2983_chan_read(struct ltc2983_desc *device, const int chan, float *val)
```
Instead of float, use a scaled integer

CiprianRegus commented on drivers/temperature/ltc2983/ltc2983.c:
```diff
@@ -0,0 +1,644 @@
+	if (ret)
+		return ret;
+
+	*val = (raw_array[3] << 24) + (raw_array[4] << 16) +
+	       (raw_array[5] << 8) + raw_array[6];
```
`no_os_get_unaligned_be32()`

CiprianRegus commented on drivers/temperature/ltc2983/ltc2983.c:
```diff
@@ -0,0 +1,644 @@
+	uint32_t reg = LTC2983_CHAN_START_ADDR(sensor->chan);
+
+	chan_val |= LTC2983_CHAN_TYPE(sensor->type);
+	uint8_t raw_array[] = {LTC2983_SPI_WRITE_BYTE, (reg >> 8) & 0xFF,
+			       reg & 0xFF, (chan_val >> 24) & 0xFF,
+			       (chan_val >> 16) & 0xFF, (chan_val >> 8) & 0xFF,
+			       chan_val & 0xFF
```
Some of these fields may be set using `no_os_put_unaligned_be32()`

CiprianRegus commented on drivers/temperature/ltc2983/ltc2983.c:
```diff
@@ -0,0 +1,644 @@
+ * @param chan_val - channel assignment value to be written
+ * @return 0 in case of success, errno errors otherwise
+ */
+int __ltc2983_chan_custom_sensor_assign(struct ltc2983_desc *device,
```
Are these functions supposed to be exported to a user? I'm asking because of the `__*` naming. If not, please make them `static`.

jemfgeronimo commented:
V2
- Changed some if statements to switch-case statements
- Changed the reset GPIO to optional
- Updated ltc2983_reg_update_bits to include the shifting inside
- Removed use of float
- Utilized no_os_put_unaligned/get_unaligned APIs
- Declared some functions as static (functions named as __*)
- Fixed some error handling (goto gpio_err)
